v2/snmp/server.go

```diff
@@ -0,0 +1,125 @@
+package snmp
+
+import (
+	"io"
+	"net"
+
+	"github.com/pkg/errors"
+
+	"github.com/geoffgarside/ber"
+)
+
+// Server provides an interface for receiving Trap and Inform messages.
+// This is only defined because it will facilitate unit testing of calling code that might want to mock the server
+// factory.
+type Server io.Closer
+
+// Handler is the interface that needs to be supported by the callback provided when a server is instantiated.
+type Handler interface {
+	// NewMessage is called when a trap/inform message has been received.
+	// pdu defines the content of the message.
+	// isInform defines the message type.
+	// sourceAddr is the address which originated the message
+	// Note that a NewMessage invocation will block the receipt of other messages.
+	// In the case of an inform message, it will also block the transmission of the acknowledgement message.
+	// It is the responsibility of the Handler implementation to return in a timely fashion.
```

Comment:
do you have an example that shows the Handler impl processing this in a Go routine?

Reply:
Only example is this (commented out at the bottom of server_test.go):
```
type handler struct {
	wg  *sync.WaitGroup
	pdu *PDU
}

func newHandler() *handler {
	return &handler{wg: &sync.WaitGroup{}}
}

func (h *handler) NewMessage(pdu *PDU, isInform bool, addr net.Addr) {
	h.pdu = pdu
	h.wg.Done()
}

func TestRealServer(t *testing.T) {

	h := newHandler()
	h.wg.Add(1)
	m, err := NewServerFactory().NewServer(context.Background(), h, Hooks(DiagnosticServerHooks))
	assert.NoError(t, err)
	defer m.Close()

	//	Generate an inform message:
	//    	snmpinform -d -v 2c -c public localhost '' 1.3.6.1.1.2.3 1.3.6.1.7.8.9 i 123456

	h.wg.Wait()
	assert.NotZero(t, h.pdu.VarbindList[0].TypedValue.Value, "upTime should be defined")
	assert.Equal(t, "1.3.6.1.1.2.3", h.pdu.VarbindList[1].TypedValue.String())
	assert.Equal(t, "123456", h.pdu.VarbindList[2].TypedValue.String())
}
```
I thought about converting this into an executable Example, but that would require extending the snmp 'client' session to implement an snmptrap method.

Reply:
ok, commented out looks good.